### PR TITLE
Update to Mirage 3 APIs

### PR DESCRIPTION
**Pull request by djs55**
Supersedes #28

### CODE REVIEW

File: lib/client.ml [hunk 1]
```diff
@@ -101,16 +109,16 @@ let list channel =
   channel.read buf
   >>= fun () ->
   match Announcement.unmarshal buf with
-  | `Error e -> fail e
-  | `Ok kind ->
+  | Result.Error e -> fail e
```
Comment (samoht) [inline lib/client.ml]: it is better to `open Result` at the beginning of the file (and this is not even needed for OCaml>4.03)

File: lib/client.ml [hunk 2]
```diff
@@ -72,15 +72,23 @@ end
 
 module Rpc = Mux.Make(NbdRpc)
 
-type info = {
-  read_write: bool;
-  sector_size: int;
-  size_sectors: int64;
-}
+(* samoht: `Msg should be the list of all possible exceptions *)
+type error = [ Mirage_block.error | `Msg of string ]
```
Comment (samoht) [inline lib/client.ml]: Ideally the `Msg of string` should be a more concrete error, e.g. something like `Protocol of Protocol.Error.t` or similar, and the `pp_error` will do the conversion to a string.

**Comment by lindig** [comment]
The CI fails on OCaml 4.02 and passes on OCaml 4.03. If this is as expected, we would be only to merge this once we are on OCaml 4.03 (together with updating most of our upstream libraries).

**Comment by mseri** [comment]
It should be enough to add JaneStreet Result wrapper as a dependency

**Comment by djs55** [comment]
I think Mirage 3 packages are incompatible with OCaml 4.02, for example:
```
$ opam upgrade mirage-types-lwt
The following dependencies couldn't be met:
  - mirage-types-lwt -> mirage-types >= 3.0.0 -> mirage-protocols >= 1.0.0
Your request can't be satisfied:
  - mirage-protocols>=1.0.0 is not available because your system doesn't comply with ocaml-version >= "4.03.0".
```
I'll update the travis and opam files to match. Out of curiosity, do you plan to upgrade to 4.03?

**Comment by avsm** [comment]
Confirmed: Mirage3 libraries are 4.03 only, although they could be backported to 4.02 with some effort.

**Comment by robhoes** [comment]
4.03? 4.04!

**Comment by jonludlam** [comment]
I think this should also have a patch similar to that in #35 so that we can use the results of read/write in other contexts.

**Comment by djs55** [comment]
PR updated.

@jonludlam I think `page_aligned_buffer = Cstruct.t` falls out of using [Mirage_block_lwt.S](https://github.com/mirage/mirage-block/blob/9d3f4f287fa348bfdebaa56f97c4242f84a10ba3/lwt/mirage_block_lwt.mli#L24) which is the new name of `V1_LWT.BLOCK` (a recent-ish Mirage 3 change)

**Comment by mseri** [comment]
I don't see other reasons not to merge it. @jonludlam what do you think?

**Comment by jonludlam** [comment]
OK, now we've got the memory leak fixed and released, we can merge this!

**Comment by djs55** [comment]
@jonludlam shall I release this (v3.0.0?) and add it to the Mirage 3 beta?

**Comment by djs55** [comment]
@jonludlam https://github.com/xapi-project/nbd/pull/39

**Pull request merged**